### PR TITLE
Remove LocalFunctionTypeParameterSymbol

### LINKED ISSUE #17244: LocalFunctionTypeParameterSymbol duplicates a lot of code
There are few differences after the lazy evaluation problems were fixed in https://github.com/dotnet/roslyn/pull/17014. We should try to unify this type with the other TypeParameterSymbols.

### PR DESCRIPTION
**Pull request by agocke**
LocalFunctionTypeParameterSymbol duplicates a lot of code from SourceMethodTypeParamaterSymbol, mainly because LocalFunctionSymbol inherits from MethodSymbol, not SourceMethodSymbol.

This PR refactors LocalFunctionSymbol to inherit from SourceMethodSymbol and removes LocalFunctionTypeParameterSymbol entirely.

Fixes #17244

### CODE REVIEW

**Comment by agocke** [comment]
@dotnet/roslyn-compielr for review. The third commit is the only really important one. The rest of the commits are just mechanical translations by the IDE to use newer language features.

**Comment by agocke** [comment]
@khyperia I think some additional refactorings that I made earlier may have helped here. And I did have to move some more logic around constraints into SourceMethodSymbol, but it seems worth it to reduce the overall complexity.

**Comment by agocke** [comment]
test windows_eta_open_prtest please

File: src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs [hunk 1]
```diff
@@ -36,14 +35,18 @@ internal sealed class LocalFunctionSymbol : MethodSymbol
             Binder binder,
             Symbol containingSymbol,
             LocalFunctionStatementSyntax syntax)
+            : base(containingSymbol, syntax.GetReference(), syntax.ExpressionBody?.GetReference() ?? syntax.Body?.GetReference(), syntax.Location)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs]: >syntax.ExpressionBody?.GetReference() ?? syntax.Body?.GetReference() [](start = 60, length = 68)

This looks inconsistent with other functions where Body is preferred over ExpressionBody.

File: src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs [hunk 2]
```diff
@@ -36,14 +35,18 @@ internal sealed class LocalFunctionSymbol : MethodSymbol
...
-            _declarationModifiers =
-                DeclarationModifiers.Private |
-                DeclarationModifiers.Static |
-                syntax.Modifiers.ToDeclarationModifiers();
+            MakeFlags(
+                methodKind: MethodKind.LocalFunction,
+                declarationModifiers: DeclarationModifiers.Private |
+                                      DeclarationModifiers.Static |
+                                      syntax.Modifiers.ToDeclarationModifiers(),
+                returnsVoid: false, /* arbitrary, flags.ReturnsVoid is unused in local functions */
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs]: > returnsVoid: false, /* arbitrary, flags.ReturnsVoid is unused in local functions */ [](start = 15, length = 84)

This doesn't feel right. This approach is fragile in the long run. Also, I believe ReturnsVoid is a public API, therefore, it is used, and it should return correct value.

File: src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs [hunk 3]
```diff
@@ -289,49 +278,14 @@ internal override TypeSymbol IteratorElementType
...
-        public override bool IsExtern => (_declarationModifiers & DeclarationModifiers.Extern) != 0;
-
-        public bool IsUnsafe => (_declarationModifiers & DeclarationModifiers.Unsafe) != 0;
-
-        internal bool IsExpressionBodied => _syntax.Body == null && _syntax.ExpressionBody != null;
-
-        public override DllImportData GetDllImportData() => null;
-
-        internal override ImmutableArray<string> GetAppliedConditionalSymbols() => ImmutableArray<string>.Empty;
+        internal override bool IsExpressionBodied => _syntax.Body == null && _syntax.ExpressionBody != null;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs]: >internal override bool IsExpressionBodied => _syntax.Body == null && _syntax.ExpressionBody != null; [](start = 8, length = 100)

It looks like there is an inconsistency between this code and the information passed in constructor initializer above.

File: src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs [hunk 4]
```diff
@@ -352,9 +306,9 @@ private static void ReportAttributesDisallowed(SyntaxList<AttributeListSyntax> a
             }
         }
 
-        private ImmutableArray<LocalFunctionTypeParameterSymbol> MakeTypeParameters(DiagnosticBag diagnostics)
+        private ImmutableArray<SourceMethodTypeParameterSymbol> MakeTypeParameters(DiagnosticBag diagnostics)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs]: >private ImmutableArray<SourceMethodTypeParameterSymbol> MakeTypeParameters(DiagnosticBag diagnostics) [](start = 8, length = 101)

Can we share some code with SourceMemberMethodSymbol.MakeTypeParameters?

File: src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs [hunk 5]
```diff
@@ -400,64 +354,62 @@ private ImmutableArray<LocalFunctionTypeParameterSymbol> MakeTypeParameters(Diag
...
-            return (clause != null) ? clause.ConstraintTypes : ImmutableArray<TypeSymbol>.Empty;
+            get
+            {
+                MakeTypeParameterConstraints();
+                return _lazyTypeParameterConstraints;
+            }
         }
 
-        private TypeParameterConstraintClause GetTypeParameterConstraintClause(int ordinal)
+        private void MakeTypeParameterConstraints()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs]: >MakeTypeParameterConstraints [](start = 21, length = 28)

Is it possible to share code with SourceMemberMethodSymbol?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs [hunk 6]
```diff
@@ -332,63 +308,21 @@ protected virtual void LazyAsyncMethodChecks(CancellationToken cancellationToken
...
+        public sealed override Symbol ContainingSymbol => _containingSymbol;
 
-        public override NamedTypeSymbol ContainingType
-        {
-            get
-            {
-                return _containingType;
-            }
-        }
+        public override NamedTypeSymbol ContainingType => (_containingSymbol as NamedTypeSymbol) ?? base.ContainingType;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: >public override NamedTypeSymbol ContainingType => (_containingSymbol as NamedTypeSymbol) ?? base.ContainingType; [](start = 8, length = 112)

Is there a good reason to override this property at all?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs [hunk 7]
```diff
@@ -535,57 +396,21 @@ internal sealed override Cci.CallingConvention CallingConvention
...
+        public override ImmutableArray<Location> Locations => this.locations;
 
-        public sealed override ImmutableArray<SyntaxReference> DeclaringSyntaxReferences
-        {
-            get
-            {
-                return (this.syntaxReferenceOpt == null) ? ImmutableArray<SyntaxReference>.Empty : ImmutableArray.Create(this.syntaxReferenceOpt);
-            }
-        }
+        public sealed override ImmutableArray<SyntaxReference> DeclaringSyntaxReferences => (this.syntaxReferenceOpt == null) ? ImmutableArray<SyntaxReference>.Empty : ImmutableArray.Create(this.syntaxReferenceOpt);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: >public sealed override ImmutableArray<SyntaxReference> DeclaringSyntaxReferences => (this.syntaxReferenceOpt == null) ? ImmutableArray<SyntaxReference>.Empty : ImmutableArray.Create(this.syntaxReferenceOpt); [](start = 8, length = 207)

This line is too long now, please break it in two.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs [hunk 8]
```diff
@@ -594,39 +419,35 @@ public override string GetDocumentationCommentXml(CultureInfo preferredCulture =
...
 
-        public sealed override int Arity
+        private TypeParameterConstraintClause GetTypeParameterConstraintClause(int ordinal)
         {
-            get
-            {
-                return TypeParameters.Length;
-            }
+            var clauses = TypeParameterConstraints;
+            return (clauses.Length > 0) ? clauses[ordinal] : null;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: >return (clauses.Length > 0) ? clauses[ordinal] : null; [](start = 12, length = 54)

Do we consider it valid to call this method with an ordinal out of range? Why?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: Same question for helpers above.

---
In reply to: [110970213](https://github.com/dotnet/roslyn/pull/18551#discussion_r110970213) [](ancestors = 110970213)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs [hunk 9]
```diff
@@ -594,39 +419,35 @@ public override string GetDocumentationCommentXml(CultureInfo preferredCulture =
...
         }
 
-        internal sealed override bool TryGetThisParameter(out ParameterSymbol thisParameter)
+        protected virtual ImmutableArray<TypeParameterConstraintClause> TypeParameterConstraints => throw ExceptionUtilities.Unreachable;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: >protected virtual ImmutableArray<TypeParameterConstraintClause> TypeParameterConstraints => throw ExceptionUtilities.Unreachable; [](start = 8, length = 129)

It feels like this method should be abstract. It should be clear from this PR how each derived class implements it.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs [hunk 10]
```diff
@@ -675,11 +490,7 @@ internal sealed override OverriddenOrHiddenMembersResult OverriddenOrHiddenMembe
             }
         }
 
-        internal sealed override bool RequiresCompletion
-        {
-            get { return true; }
-        }
-
+        internal sealed override bool RequiresCompletion => true;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: >internal sealed override bool RequiresCompletion => true; [](start = 8, length = 57)

Is this implementation correct for local functions?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs [hunk 11]
```diff
@@ -1260,7 +1053,7 @@ private void DecodeDllImportAttribute(ref DecodeWellKnownAttributeArguments<Attr
                 hasErrors = true;
             }
 
-            if (this.IsGenericMethod || (object)_containingType != null && _containingType.IsGenericType)
+            if (this.IsGenericMethod || ContainingType?.IsGenericType == true)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: >ContainingType [](start = 40, length = 14)

Please revert back to ```_containingType```.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs [hunk 12]
```diff
@@ -1390,7 +1183,7 @@ internal override void PostDecodeWellKnownAttributes(ImmutableArray<CSharpAttrib
                             if (!this.IsAbstract && !this.IsExtern)
                             {
                                 // CS0423: Since '{1}' has the ComImport attribute, '{0}' must be extern or abstract
-                                diagnostics.Add(ErrorCode.ERR_ComImportWithImpl, this.Locations[0], this, _containingType);
+                                diagnostics.Add(ErrorCode.ERR_ComImportWithImpl, this.Locations[0], this, _containingSymbol);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: >_containingSymbol [](start = 106, length = 17)

This is not necessary a type now.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs [hunk 13]
```diff
@@ -1372,7 +1165,7 @@ internal override void PostDecodeWellKnownAttributes(ImmutableArray<CSharpAttrib
                 Debug.Assert(_lazyCustomAttributesBag != null);
                 Debug.Assert(_lazyCustomAttributesBag.IsDecodedWellKnownAttributeDataComputed);
 
-                if (_containingType.IsComImport && _containingType.TypeKind == TypeKind.Class)
+                if (ContainingType.IsComImport && ContainingType.TypeKind == TypeKind.Class)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbol.cs]: Consider caching and reusing ```ContainingType```. This code is valid only for member methods, not for local functions.

**Comment by AlekseyTs** [comment]
Done with review pass.

File: src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs [hunk 14]
```diff
@@ -400,64 +354,62 @@ private ImmutableArray<LocalFunctionTypeParameterSymbol> MakeTypeParameters(Diag
...
-            if (_lazyTypeParameterConstraints == null)
+            if (!_lazyTypeParameterConstraints.IsDefault)
             {
-                var diagnostics = DiagnosticBag.GetInstance();
-                var constraints = MakeTypeParameterConstraints(diagnostics);
+                return;
+            }
 
-                lock (_declarationDiagnostics)
+            var diagnostics = DiagnosticBag.GetInstance();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/LocalFunctionSymbol.cs]: `diagnostics` is not used.

**Comment by jaredpar** [comment]
@agocke looks like this PR is getting stale.  Will we be able to get it done soon?  Don't think this will meet the bar next week.

**Comment by agocke** [comment]
@jaredpar I thought the bar had already passed. I was waiting for a new branch to open. If it's still possible to get this in, I'll fix it up and get it in today.

**Comment by jaredpar** [comment]
It's simplifying code in a place that has caused us a few issues.  Still meets the bar for me.

**Comment by agocke** [comment]
Closing this PR as it doesn't currently meet the bar and needs some more refactoring work to break out a new subclass and unify it with `LambdaMethodSymbol`

**Pull request closed without merging**